File: 85. maximal-rectangle.py
from typing import List

class Solution:
    def maximalRectangle(self, matrix: List[List[str]]) -> int:
        if not matrix:
            return 0
        
        rows, cols = len(matrix), len(matrix[0])
        heights = [0] * (cols + 1)  # Include an extra element for easier calculation
        max_area = 0
        
        for row in matrix:
            for i in range(cols):
                heights[i] = heights[i] + 1 if row[i] == '1' else 0
            
            # Calculate max area using histogram method
            stack = []
            for i in range(len(heights)):
                while stack and heights[i] < heights[stack[-1]]:
                    h = heights[stack.pop()]
                    w = i if not stack else i - stack[-1] - 1
                    max_area = max(max_area, h * w)
                stack.append(i)
        
        return max_area

# Test cases
if __name__ == "__main__":
    solution = Solution()
    
    # Test case 1
    matrix = [["1","0","1","0","0"],["1","0","1","1","1"],["1","1","1","1","1"],["1","0","0","1","0"]]
    output = 6
    result = solution.maximalRectangle(matrix)
    print(f"Test 1: {result} == {output} -> {'PASS' if result == output else 'FAIL'}")
    
    # Test case 2
    matrix = [["0"]]
    output = 0
    result = solution.maximalRectangle(matrix)
    print(f"Test 2: {result} == {output} -> {'PASS' if result == output else 'FAIL'}")
    
    # Test case 3
    matrix = [["1"]]
    output = 1
    result = solution.maximalRectangle(matrix)
    print(f"Test 3: {result} == {output} -> {'PASS' if result == output else 'FAIL'}")

File: 85. maximal-rectangle.java
import java.util.*;

class Solution {
    public int maximalRectangle(char[][] matrix) {
        if (matrix == null || matrix.length == 0) {
            return 0;
        }
        
        int rows = matrix.length;
        int cols = matrix[0].length;
        int[] heights = new int[cols + 1]; // Include an extra element for easier calculation
        int maxArea = 0;
        
        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                heights[c] = matrix[r][c] == '1' ? heights[c] + 1 : 0;
            }
            
            // Calculate max area using histogram method
            Stack<Integer> stack = new Stack<>();
            for (int i = 0; i < heights.length; i++) {
                while (!stack.isEmpty() && heights[i] < heights[stack.peek()]) {
                    int h = heights[stack.pop()];
                    int w = stack.isEmpty() ? i : i - stack.peek() - 1;
                    maxArea = Math.max(maxArea, h * w);
                }
                stack.push(i);
            }
        }
        
        return maxArea;
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        
        // Test case 1
        char[][] matrix1 = {
            {'1','0','1','0','0'},
            {'1','0','1','1','1'},
            {'1','1','1','1','1'},
            {'1','0','0','1','0'}
        };
        int output1 = 6;
        int result1 = solution.maximalRectangle(matrix1);
        System.out.println("Test 1: " + result1 + " == " + output1 + " -> " + (result1 == output1 ? "PASS" : "FAIL"));
        
        // Test case 2
        char[][] matrix2 = {{'0'}};
        int output2 = 0;
        int result2 = solution.maximalRectangle(matrix2);
        System.out.println("Test 2: " + result2 + " == " + output2 + " -> " + (result2 == output2 ? "PASS" : "FAIL"));
        
        // Test case 3
        char[][] matrix3 = {{'1'}};
        int output3 = 1;
        int result3 = solution.maximalRectangle(matrix3);
        System.out.println("Test 3: " + result3 + " == " + output3 + " -> " + (result3 == output3 ? "PASS" : "FAIL"));
    }
}
